### ISSUE TITLE
panic runtime error while unmarshal `bad-utf8-at-end.toml` on go 1.17

### ISSUE THREAD

**Issue by afdesk** [BurntSushi/toml#329]
Hi!
thanks a lot for this useful package!

but we got a problem with [BurntSushi/toml](https://github.com/BurntSushi/toml/).
There is a panic while unmarshal [bad-utf8-at-end.toml](https://github.com/BurntSushi/toml/blob/master/internal/toml-test/tests/invalid/encoding/bad-utf8-at-end.toml) from your tests on go 1.17.

it seems that `runtime.panic` was changes: [cmd/compile,runtime: provide index information on bounds check failure ](https://github.com/golang/go/commit/2c423f063b137781a97a6a56998a2e3abffc4051).
and this check doesn't work now: [https://github.com/BurntSushi/toml/blob/9be4ccf5153f55ef1032bdf52d01e65ea480380b/parse.go#L28](https://github.com/BurntSushi/toml/blob/9be4ccf5153f55ef1032bdf52d01e65ea480380b/parse.go#L28)

For reproduce I wrote next sample:
```go
	b, _ := ioutil.ReadFile("bad-utf8-at-end.toml")
	var v interface{}
	if err := toml.Unmarshal(b, &v); err != nil {
		log.Fatalf("toml.Unmarshal error: %v", err)
	}
```
Result:
```
panic: runtime error: slice bounds out of range [238:237] [recovered]
        panic: runtime error: slice bounds out of range [238:237]

goroutine 1 [running]:
github.com/BurntSushi/toml.parse.func1()
        /home/home/go/pkg/mod/github.com/!burnt!sushi/toml@v0.4.1/parse.go:46 +0x16a
panic({0x5072a0, 0xc000148018})
        /usr/local/go/src/runtime/panic.go:1047 +0x262
github.com/BurntSushi/toml.(*lexer).current(0xc000150000)
        /home/home/go/pkg/mod/github.com/!burnt!sushi/toml@v0.4.1/lex.go:129 +0x90
github.com/BurntSushi/toml.(*lexer).emit(0xc000150000, 0x6)
        /home/home/go/pkg/mod/github.com/!burnt!sushi/toml@v0.4.1/lex.go:133 +0x27
github.com/BurntSushi/toml.lexMultilineString(0xc000150000)
        /home/home/go/pkg/mod/github.com/!burnt!sushi/toml@v0.4.1/lex.go:693 +0x2f1
github.com/BurntSushi/toml.(*lexer).nextItem(0xc000150000)
        /home/home/go/pkg/mod/github.com/!burnt!sushi/toml@v0.4.1/lex.go:98 +0xf9
github.com/BurntSushi/toml.(*parser).next(0xc0001401c0)
        /home/home/go/pkg/mod/github.com/!burnt!sushi/toml@v0.4.1/parse.go:95 +0x55
github.com/BurntSushi/toml.(*parser).topLevel(0xc0001401c0, {0x12, {0xc00014c100, 0x0}, 0x5})
        /home/home/go/pkg/mod/github.com/!burnt!sushi/toml@v0.4.1/parse.go:172 +0x732
github.com/BurntSushi/toml.parse({0xc00014c100, 0xf2})
        /home/home/go/pkg/mod/github.com/!burnt!sushi/toml@v0.4.1/parse.go:79 +0x4f5
github.com/BurntSushi/toml.(*Decoder).Decode(0xc000116350, {0x4f7360, 0xc000116340})
        /home/home/go/pkg/mod/github.com/!burnt!sushi/toml@v0.4.1/decode.go:120 +0x625
github.com/BurntSushi/toml.Decode({0xc00014c000, 0xf2}, {0x4f7360, 0xc000116340})
        /home/home/go/pkg/mod/github.com/!burnt!sushi/toml@v0.4.1/decode.go:135 +0x10f
github.com/BurntSushi/toml.Unmarshal({0xc00014a000, 0xf2, 0x200}, {0x4f7360, 0xc000116340})
        /home/home/go/pkg/mod/github.com/!burnt!sushi/toml@v0.4.1/decode.go:23 +0x85
main.main()
        /home/home/t/main.go:12 +0xe5
```

**upd**: I can try to create a PR for fix it )

**Comment by arp242** [comment]
Yeah, this is a known problem; this test should be skipped though, so not sure how you're getting this error?

It's a bit of an obscure edge case, so I haven't bothered fixing it. Patches are always welcome, of course.

**Comment by afdesk** [comment]
this package is used for [trivy](https://github.com/aquasecurity/trivy), and if it scans itself we will get this problem...

**Comment by arp242** [comment]
> if it scans itself we will get this problem

I think that's a different issue than the one that `bad-utf8-at-end.toml` tests, since that's only triggered on invalid UTF-8.

I downloaded that trivy repo and ran `./trivy fs .` after building it, and that seems to work well(?) How do you trigger this problem? Or what error do you get?

**Comment by afdesk** [comment]
the problem appears for run inside docker.

there is an issue [aquasecurity/trivy/issues/1372](https://github.com/aquasecurity/trivy/issues/1372).

**Comment by arp242** [comment]
I found https://github.com/aquasecurity/trivy/issues/1372 – I assume that's the related issue. Guess it only shows with `trivy image` and not `trivy fs`(?)

Anyway, so what's going on is that it seems it tries to scan all `*.toml` files it can find, including the test files bundled in this library, one of which currently causes a panic. That explains how you're actually triggering this.

Would moving the files to a `testdata` directory or some such help? Are those directories ignored by `trivy`? Even if this panic is fixed, it's still a lot of pointless noise in that `-debug` output, and it slows things down as well (there are hundreds of files, many of which invalid).

Also, there is no guarantee this won't happen again with some new test. I think ignoring these files would be a better long-term solution than fixing this specific issue (although it obviously should be fixed).

**Comment by afdesk** [comment]
>  Guess it only shows with `trivy image` and not `trivy fs`(?)

yes, you're right!

Thanks for your help and ideas. We'll discuss about it.

**Comment by capnspacehook** [comment]
I may have stumbled across the same bug while fuzzing, does this reproduce the same thing?

https://go.dev/play/p/y7XA9_OS_fL

**Comment by arp242** [comment]
Yeah, that seems like the same issue @capnspacehook

**Comment by arp242** [comment]
Fixed in 891d261
